### PR TITLE
Added assertions to narrow down assertion in Search for RocksDBTicks

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

Fix merge conflicts and add assertions.
It doesn't fix tests with replication2 truncate and search for now, but at least make clear that issue with these tests in whole snapshot, not just counting

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp [hunk 1]
```diff
@@ -225,32 +225,25 @@ auto DocumentStateShardHandler::prepareShardsForLogReplay() noexcept -> void {
...
-            .commit(true);
+        auto& idx =
+            basics::downCast<iresearch::IResearchRocksDBInvertedIndex>(*index);
+        TRI_ASSERT(!idx._isCreation) << "Inverted index still in creation mode";
+        auto res = idx.commit(true);
+        TRI_ASSERT(res.ok()) << "Failed to do Inverted index commit";
+        TRI_ASSERT(
+            res.get() ==
+            arangodb::iresearch::IResearchDataStore::CommitResult::NO_CHANGES)
+            << "Inverted index still has changes after commit.";
```
Comment (MBkkt) [inline arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:236]:
```suggestion
        auto res = idx.commit(true);
#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
        TRI_ASSERT(res.ok()) << "Failed to do first Inverted index commit";
        res = idx.commit(true);
        TRI_ASSERT(res.ok()) << "Failed to do second Inverted index commit";
        TRI_ASSERT(
            res.get() ==
            arangodb::iresearch::IResearchDataStore::CommitResult::NO_CHANGES)
            << "Inverted index still has changes after first commit.";
#endif
```

Same for link

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js [hunk 2]
```diff
@@ -89,8 +94,11 @@ function recoverySuite () {
       var p = v.properties().links;
       assertTrue(p.hasOwnProperty(cn));
       var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js:97]: NIT: We can check with explain, that actually no-materialize used

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js [hunk 3]
```diff
@@ -89,8 +94,11 @@ function recoverySuite () {
...
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(result[0], expectedResult[0]);
+      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. StoredValue _key values: ${JSON.stringify(expectedResultStoredValues)}`;
+      assertEqual(result[0], expectedResult[0], message);
+      assertEqual(result[0], expectedResultStoredValues.length, message);
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js:101]: We probably want 4 queries:
1. collection
2. view count
3. view stored value
4. view materialize from rocksdb

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js [hunk 4]
```diff
@@ -89,8 +94,13 @@ function recoverySuite () {
       var p = v.properties().links;
       assertTrue(p.hasOwnProperty(cn));
       var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(result[0], expectedResult[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
+      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js:100]:
```suggestion
      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues.length} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate.js [hunk 5]
```diff
@@ -73,20 +73,19 @@ function recoverySuite () {
...
       assertEqual(v.name(), vn);
       assertEqual(v.type(), 'arangosearch');
       var p = v.properties().links;
       assertTrue(p.hasOwnProperty(cn));
       var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(result[0], expectedResult[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
+      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate.js:85]:
```suggestion
      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues.length} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate.js [hunk 6]
```diff
@@ -73,20 +73,19 @@ function recoverySuite () {
...
       var v = db._view(vn);
       assertEqual(v.name(), vn);
       assertEqual(v.type(), 'arangosearch');
       var p = v.properties().links;
       assertTrue(p.hasOwnProperty(cn));
       var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(result[0], expectedResult[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate.js:84]:
```suggestion
      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._id").toArray();
```

File: tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js [hunk 7]
```diff
@@ -89,8 +94,13 @@ function recoverySuite () {
       var p = v.properties().links;
       assertTrue(p.hasOwnProperty(cn));
       var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(result[0], expectedResult[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-arangosearch-link-populate-truncate-no-flushthread.js:99]:
```suggestion
      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._id").toArray();
```

File: tests/js/client/recovery/search/view-search-alias-populate-truncate-no-flushthread.js [hunk 8]
```diff
@@ -86,9 +79,14 @@ function recoverySuite () {
       };
       checkView(vn, "i1");
 
-      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(expectedResult[0], result[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
+      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-search-alias-populate-truncate-no-flushthread.js:86]:
```suggestion
      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues.length} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```

File: tests/js/client/recovery/search/view-search-alias-populate-truncate-no-flushthread.js [hunk 9]
```diff
@@ -86,9 +79,14 @@ function recoverySuite () {
       };
       checkView(vn, "i1");
 
-      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(expectedResult[0], result[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-search-alias-populate-truncate-no-flushthread.js:85]:
```suggestion
      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._id").toArray();
```

File: tests/js/client/recovery/search/view-search-alias-populate-truncate.js [hunk 10]
```diff
@@ -84,9 +77,14 @@ function recoverySuite () {
       };
       checkView(vn, "i1");
 
-      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(expectedResult[0], result[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
+      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-search-alias-populate-truncate.js:84]:
```suggestion
      const message = `Search with COUNT ${result[0]}, Collection COUNT ${expectedResult[0]}, SearchStoredValues: ${expectedResultStoredValues.length}. Search Not StoradeValues: ${expectedResultNonStoredValues.length} StoredValue _key values: ${JSON.stringify(expectedResultStoredValues.sort())}, not stored: ${JSON.stringify(expectedResultNonStoredValues.sort())}`;
```

File: tests/js/client/recovery/search/view-search-alias-populate-truncate.js [hunk 11]
```diff
@@ -84,9 +77,14 @@ function recoverySuite () {
       };
       checkView(vn, "i1");
 
-      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var result = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} COLLECT WITH COUNT INTO length RETURN length").toArray();
+      var expectedResultStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
       var expectedResult = db._query("FOR doc IN " + cn + " FILTER doc.c >= 0 COLLECT WITH COUNT INTO length RETURN length").toArray();
-      assertEqual(expectedResult[0], result[0]);
+      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._key").toArray();
```
Comment (MBkkt) [inline tests/js/client/recovery/search/view-search-alias-populate-truncate.js:83]:
```suggestion
      var expectedResultNonStoredValues = db._query("FOR doc IN " + vn + " SEARCH doc.c >= 0 OPTIONS {waitForSync: true} RETURN doc._id").toArray();
```

**Review by MBkkt — APPROVED** [review]

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp [hunk 12]
```diff
@@ -226,32 +226,31 @@ auto DocumentStateShardHandler::prepareShardsForLogReplay() noexcept -> void {
...
       if (index->type() == Index::TRI_IDX_TYPE_INVERTED_INDEX) {
-        basics::downCast<iresearch::IResearchRocksDBInvertedIndex>(*index)
-            .commit(true);
+        auto& idx =
+            basics::downCast<iresearch::IResearchRocksDBInvertedIndex>(*index);
+        TRI_ASSERT(!idx._isCreation) << "Inverted index still in creation mode";
+        auto res = idx.commit(true);
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+        TRI_ASSERT(res.ok()) << "Failed to do first Inverted index commit";
+        res = idx.commit(true);
```
Comment (jsteemann) [inline arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:235]: Is it actually intentional that we do two commits here in maintainer mode? can that have any side effects?

Comment (MBkkt) [reply arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:235]: Yep, we do second commit, and check that nothing changes

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**